README.md
<p align="center">
  <img alt="gitlab-runner" src="https://i.stack.imgur.com/skYjT.png" width="500px" float="center"/>
</p>
<h1 align="center">💻 Windows GitLab Runner Setup 💻</h1>
<p align="center">
  <strong>Passo a passo para criação de um Runner para execução de projetos Windows</strong>
</p>

## Description

Setup com todos os pacotes necessários para configurar um Runner no Windows baseado em uma EC2 com Containers.

## Add to git and push

You must send the project to your GitHub after the modifications

```bash
git add -f .
git commit -m "Added - Fixing somethings"
git push origin master
```

## 📋 Versioning

- [CHANGELOG](CHANGELOG.md)

## Project Status

* 🔛 In production


CHANGELOG.md
# CHANGELOG

All important changes to this project will be added to this file! This changelog will be based on [Keep a change log](http://keepachangelog.com/)

## 1.0.0 - [ 01/04/2020 ]

### Added

* Initial commit with basic project structure.
* Including REAMDE.md and CHANGELOG.md files in their first versions.
* Create PowerShell script to automate Chocolatey instalation with gitlab-runner and git.
* AWS configure script in PowerShell to install AWS CLI MSI.
* Adding Admin flux in PowerShell execution script.
* Validation to check if choco command and your folder exist in the system.

## 1.0.1 - [ 02/04/2020 ]

### Added

* Create Functions to each action:
  * Write-Header()
  * Log()
  * Get-Admin-Execution()
  * Get-Choco-Setup()
  * Get-AWSCLI-Setup()
  * Get-Choco-Installation()
  * Get-AWSCLI-Installation()
  * Test-Choco()
  * Test-AWSCLI()
* Improving the script execution flow.
* Fixing some erros in Setup script.
* Adding CLI params with switch case flow. Commands:
  * all (default command)
  * choco
  * aws
  * list
  * update
  * hel
* Create install script to put Get-Setup.ps1 inside your PC.
* Try-Catch flow to choco and aws.

## 1.0.1 - [ 04/04/2020 ]

### Added

* Install Runner by script.
* Adding new Command in CLI.
* Fixing Installer Script and create new Functions:
  * Write-Header()
  * Log()
  * Get-Install()
* Dockerfile with PowerShell.
* Basic Docker-Compose with .env strategy.
* Makefile with basic commands.


setup.ps1
# Test-Admin is not available yet, so use...
if (!([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole] "Administrator")) {
    Start-Process powershell -ArgumentList "-noprofile -NoExit -file `"$PSCommandPath`"" -Verb RunAs
    Exit
}

# From a Administrator PowerShell, if Get-ExecutionPolicy returns Restricted, run:
if ((Get-ExecutionPolicy) -eq "Restricted") {
    Set-ExecutionPolicy Unrestricted -Force
}

# Validate Choco command.
$ChocoInstalled = $false
if (Get-Command choco.exe -ErrorAction SilentlyContinue) {
  Write-Host "`n`nChocolatey command Alredy in the System."
  $ChocoInstalled = $true
} else {
  # Validate path to Install chocolatey.
  if (!(Test-Path 'C:\ProgramData\chocolatey\bin\choco.exe')) {
    $URL_CHOCOLATEY = "https://chocolatey.org/install.ps1"
    Write-Host "`n`nInstalling Chocolatey...`n`n"
    [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072; iex ((New-Object System.Net.WebClient).DownloadString($URL_CHOCOLATEY))
  }else{
    Write-Host "`n`nChocolatey Alredy in the System."
  }
}

# Declaring all packages that we install.
$PACKAGES = @"
gitlab-runner
"@

# Multiline text -> array.
$PACKAGES = $PACKAGES -Split [System.Environment]::NewLine

# Trim each text package.
foreach ($PACKAGE in $PACKAGES) {
  $PACKAGE = $PACKAGE | % { $_.Trim() }
}

Write-Host "`n`nInstalling packages from Chocolatey.`n`n"

# Install pakcages
foreach ($PACKAGE in $PACKAGES) {
  choco install --confirm $PACKAGE
}

Write-Host "`n`nInstalling Git.`n`n"
choco install git -params '"/GitAndUnixToolsOnPath"'

Write-Host "`n`nUpgrading all packages from Chocolatey.`n`n"
choco upgrade all -y

Write-Host "`n`nList all installed packages.`n`n"

choco list -li

Write-Host "`n`nFinish Choco installation and configuration..."

# Validate if AWSCLI64PY3.msi alredy in the system.
if (!(Test-Path 'C:\AWSCLI64PY3.msi')) {
  $AWSCLI = "https://s3.amazonaws.com/aws-cli/AWSCLI64PY3.msi"
  $PATH = Join-Path C:\ (Split-Path $AWSCLI -Leaf)
  Write-Host "`n`nInstalling AWS CLI.`n`n"
  Invoke-WebRequest $AWSCLI -OutFile $PATH
}else{
  Write-Host "`n`nThe file AWSCLI64PY3.msi Alredy in the System."
  # Validate AWS Command.
  $AwsInstalled = $false
  if (Get-Command aws -ErrorAction SilentlyContinue) {
    $AwsInstalled = $true
    Write-Host "`n`nAWS command Alredy in the System.`n`n"
    # Setting AWS Variables
    [Environment]::SetEnvironmentVariable("AWS_ACCESS_KEY_ID", "AWS_ACCESS_KEY_ID", "Machine")
    [Environment]::SetEnvironmentVariable("AWS_SECRET_ACCESS_KEY", "AWS_SECRET_ACCESS_KEY", "Machine")
    [Environment]::SetEnvironmentVariable("AWS_DEFAULT_REGION", "AWS_DEFAULT_REGION", "Machine")
    [Environment]::SetEnvironmentVariable("AWS_DEFAULT_OUTPUT", "AWS_DEFAULT_OUTPUT", "Machine")
    # To create a ~\.aws\credentials file run.
    # Default output format [None]: JSON (json) , Tab-delimited text (text) , ASCII-formatted table (table)
    aws configure --profile [name]
    aws s3 ls
  } else {
    Write-Host "`n`nError - AWS CLI is not in the System.`n`n"
  }
}

Write-Host "`n`nFinish AWS installation and configuration..."


code/Get-Install.ps1
################################################################################
## DESCRIPTION: Installation Script.
## NAME: Get-Install.ps1
## AUTHOR: Lucca Pessoa da Silva Matos
## DATE: 04.04.2020
## VERSION: 1.1
## EXEMPLE:
##     PS C:\> .\Get-Install.ps1
################################################################################

# ******************************************************************************
# FUNCTIONS
# ******************************************************************************

Function Write-Header {
  Write-Host ""
  Write-Host "========================================" -ForegroundColor Green
  Write-Host "= Install Setup GitLab Runner Script" -ForegroundColor Green
  Write-Host "= "
  Write-Host "= Author: Lucca Pessoa" -ForegroundColor Yellow
  Write-Host "= Date: 04-04-2020" -ForegroundColor Yellow
  Write-Host "= Version: 1.1" -ForegroundColor Yellow
  Write-Host "========================================" -ForegroundColor Green
  Write-Host "`n"
} #End Write-Header

Function Log($MESSAGE){
  Write-Host
  Write-Host -ForegroundColor Yellow -BackgroundColor Black $MESSAGE
}#End Log

Function Get-Install {
  $SETUP_URL="https://raw.githubusercontent.com/lpmatos/aws-gitlab-runner-windows/master/Get-Setup.ps1"
  $PATH = Join-Path C:\ (Split-Path $SETUP_URL -Leaf)
  Log("Install Script...")
  Invoke-WebRequest $SETUP_URL -OutFile $PATH
}#End Get-Install

# ******************************************************************************
# MAIN
# ******************************************************************************

Write-Header

If (!(Test-Path "C:\Get-Setup.ps1")){
  Get-Install
}
Else {
  Log("Get-Setup.ps1 alredy in the system...")
}


code/Get-Refresh.ps1
################################################################################
## DESCRIPTION: Refresh Environments Script.
## NAME: Get-Refresh.ps1 
## AUTHOR: Lucca Pessoa da Silva Matos 
## DATE: 04.04.2020
## VERSION: 1.1
## EXEMPLE: 
##     PS C:\> .\Get-Refresh.ps1
################################################################################

# ******************************************************************************
# FUNCTIONS
# ******************************************************************************

Function Write-Header {
  Write-Host ""
  Write-Host "========================================" -ForegroundColor Green
  Write-Host "= Refresh Environments in PowerShell" -ForegroundColor Green
  Write-Host "= "
  Write-Host "= Author: Lucca Pessoa" -ForegroundColor Yellow
  Write-Host "= Date: 04-04-2020" -ForegroundColor Yellow
  Write-Host "= Version: 1.1" -ForegroundColor Yellow
  Write-Host "========================================" -ForegroundColor Green
  Write-Host "`n"
} #End Write-Header

Function Log($MESSAGE){
  Write-Host
  Write-Host -ForegroundColor Yellow -BackgroundColor Black $MESSAGE
}#End Log

function Get-Refresh-Path {
  $env:Path = [System.Environment]::GetEnvironmentVariable("Path","Machine") +
    ";" + [System.Environment]::GetEnvironmentVariable("Path","User")
}#End Get-Refresh-Path

# ******************************************************************************
# MAIN
# ******************************************************************************

Write-Header

Log("Refresh Environment Variables in PowerShell...")

Get-Refresh-Path


code/Get-Setup.ps1
################################################################################
## DESCRIPTION: Setup used to configure a GitLab Runner on Windows EC2 - AWS
##              with AWS CLI, GIT and others packages.
## NAME: Get-Setup.ps1
## AUTHOR: Lucca Pessoa da Silva Matos
## DATE: 04.04.2020
## VERSION: 1.1
## EXEMPLE:
##     PS C:\> .\Get-Setup.ps1
################################################################################

[CmdletBinding()]
Param(
  [Parameter(HelpMessage="AWS EC2 GitLab Runner - Setup CLI commands.")]
  [ValidateSet("all", "choco", "aws", "register", "unregister", "runners", "list", "update", "refresh", "help")]
  [string]$Setup="all"
)

$AWSAccessKey="YourAWSAccessKey" # Replace with your access key
$AWSSecretKey="YourAWSSecretKey" # Replace with your secret key

$GitLabHost="YourGitLabHost" # Replace with your gitlab host
$GitLabToken="YourGitLabToken" # Replace with your gitlab token

$RunnerDescription="aws-gitlab-runner-ec2-windows" # Replace with your runner description
$RunnerTags="windows" # Replace with your runner tags

# ******************************************************************************
# FUNCTIONS
# ******************************************************************************

Function Write-Header {
  Write-Host ""
  Write-Host "========================================" -ForegroundColor Green
  Write-Host "= Setup AWS EC2 GitLab Runner" -ForegroundColor Green
  Write-Host "= Chocolatey, Git, AWS CLI and GitLab Runner" -ForegroundColor Green
  Write-Host "= "
  Write-Host "= Author: Lucca Pessoa" -ForegroundColor Yellow
  Write-Host "= Date: 04-04-2020" -ForegroundColor Yellow
  Write-Host "= Version: 1.1" -ForegroundColor Yellow
  Write-Host "========================================" -ForegroundColor Green
  Write-Host "`n"
} #End Write-Header

Function Log($MESSAGE){
  Write-Host
  Write-Host -ForegroundColor Yellow -BackgroundColor Black $MESSAGE
}#End Log

Function Get-Admin-Execution {
  # Test-Admin is not available yet, so use...
  If (!([Security.Principal.WindowsPrincipal][Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole] "Administrator")) {
    Start-Process powershell -ArgumentList "-noprofile -NoExit -file `"$PSCommandPath`"" -Verb RunAs
    Exit
  }
  # From a Administrator PowerShell, if Get-ExecutionPolicy returns Restricted, run:
  If ((Get-ExecutionPolicy) -eq "Restricted") {
    Set-ExecutionPolicy Unrestricted -Force
  }
} #End Get-Admin-Execution

Function Get-Choco-Setup {
  $PACKAGES = "gitlab-runner",
  "python",
  "pip"
  Log("Installing packages...")
  foreach ($PACKAGE in $PACKAGES){
    choco install --confirm $PACKAGE
  }
  Log("Installing Git...")
  choco install git -params '"/GitAndUnixToolsOnPath"'
}#End Get-Choco-Setup

Function Get-AWSCLI-Setup {
  # Setting AWS Variables
  If ($AWSAccessKey -and $AWSSecretKey) {
    Log("AWS credentials have been defined. Setting credentials...")
    aws configure set AWS_ACCESS_KEY_ID $AWSAccessKey
    aws configure set AWS_SECRET_ACCESS_KEY $AWSSecretKey
    aws configure set default.region us-east-1
    try{
      Log("Configure AWS Profile.");
      aws configure --profile [name]
      aws s3 ls
    }
    catch{
      Write-Error "Error - Configure AWS Profile..."
    }
  }
  Else {
    Write-Error "Error - AWS credentials not been defined. Bye Bye :)"
    Exit
  }
} #End Get-AWSCLI-Setup

Function Get-Choco-Installation {
  $URL_CHOCOLATEY = "https://chocolatey.org/install.ps1"
  [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072
  Log("Installing Choco...")
  Invoke-Expression ((New-Object System.Net.WebClient).DownloadString($URL_CHOCOLATEY))
} #End Get-Choco-Installation

Function Get-AWSCLI-Installation {
  $AWSCLI_URL = "https://s3.amazonaws.com/aws-cli/AWSCLI64PY3.msi"
  $PATH = Join-Path C:\ (Split-Path $AWSCLI_URL -Leaf)
  Log("Installing AWSCLI .msi...")
  Invoke-WebRequest $AWSCLI_URL -OutFile $PATH
} #End Get-AWSCLI-Installation

Function Test-Choco {
  # Validate Chocolatey Environment and Path.
  try {
    If ($env:ChocolateyInstall -or (Test-Path "$env:ChocolateyInstall") -or (Get-Command choco.exe -ErrorAction SilentlyContinue)){
      Log("Chocolatey is already installed in the system.")
      Get-Choco-Setup
    }
  }
  catch {
    Log("Chocolatey ins't in the System...")
    Get-Choco-Installation
    Get-Choco-Setup
  }
} #End Test-Choco

Function Test-AWSCLI {
  If (Get-Command aws -ErrorAction SilentlyContinue){
    Log("AWSCLI is already in the system.")
    Get-AWSCLI-Setup
  }
  Else {
    Write-Error "Error - AWSCLI ins't in the System..."
    If (!(Test-Path "C:\AWSCLI64PY3.msi")){
      Get-AWSCLI-Installation
    }
    Else {
      Log("AWSCLI .msi alredy in the system...")
    }
  }
}#End Test-AWSCLI

Function Get-GitLab-Runner-Install {
  If (Get-Command gitlab-runner -ErrorAction SilentlyContinue){
    Log("Setup GitLab Runner - Status...")
    gitlab-runner status
    Log("Setup GitLab Runner - Install...")
    gitlab-runner install
  }
  Else {
    Write-Error "Error - GitLab Runner ins't in the System..."
  }
}#End Get-GitLab-Runner-Install

Function Get-GitLab-Runner-Register {
  Get-GitLab-Runner-Install
  If ($GitLabHost -and $GitLabToken) {
    try {
      Log("Setup GitLab Runner - Register...")
      gitlab-runner register -n --url $GitLabHost --registration-token $GitLabToken --description $RunnerDescription --tag-list $RunnerTags --executor shell --shell powershell --request-concurrency 15 --run-untagged false
      Log("Setup GitLab Runner - Status...")
      gitlab-runner status
      Log("Setup GitLab Runner - Stop...")
      gitlab-runner stop
      Log("Setup GitLab Runner - Start...")
      gitlab-runner start
      Log("Setup GitLab Runner - Status...")
      gitlab-runner status
    }
    catch {
      Write-Error "Error - GitLab Runner Registration"
    }
  }
  Else {
    Write-Error "Error - GitLab Runner credentials not been defined. Bye Bye :)"
    Exit
  }
}#End Get-GitLab-Runner-Register

function Get-Refresh-Path {
  $env:Path = [System.Environment]::GetEnvironmentVariable("Path","Machine") +
    ";" + [System.Environment]::GetEnvironmentVariable("Path","User")
}#End Get-Refresh-Path

# ******************************************************************************
# MAIN
# ******************************************************************************

Write-Header

Get-Admin-Execution

switch ($Setup) {
  all {
    Test-Choco
    Test-AWSCLI
    if (Get-Command aws -ErrorAction SilentlyContinue){
      Log("AWSCLI is already in the system.")
      Get-AWSCLI-Setup
    }
    Get-GitLab-Runner-Register
  }

  choco {
    Test-Choco
  }

	aws {
    Test-AWSCLI
    if (Get-Command aws -ErrorAction SilentlyContinue){
      Log("AWSCLI is already in the system.")
      Get-AWSCLI-Setup
    } Else {
      Write-Error "Error - AWSCLI ins't in the System... Bye Bye :)"
      Exit
    }
  }

  register {
    If (Get-Command gitlab-runner -ErrorAction SilentlyContinue){
      Get-GitLab-Runner-Register
    } Else {
      Write-Error "Error - GitLab Runner ins't in the System... Bye Bye :)"
      Exit
    }
  }

  unregister {
    If (Get-Command gitlab-runner -ErrorAction SilentlyContinue){
      gitlab-runner unregister --all-runners
    }
    Else {
      Write-Error "Error - GitLab Runner ins't in the System... Bye Bye :)"
      Exit
    }
  }

  runners {
    If (Get-Command gitlab-runner -ErrorAction SilentlyContinue){
      gitlab-runner list
    }
    Else {
      Write-Error "Error - GitLab Runner ins't in the System... Bye Bye :)"
      Exit
    }
  }

	list {
    If (Get-Command choco -ErrorAction SilentlyContinue){
      Log("List all installed packages.")
		  choco list -li
    }
    Else {
      Write-Error "Error - Choc ins't in the System... Bye Bye :)"
      Exit
    }
	}

	update {
    If (Get-Command choco -ErrorAction SilentlyContinue){
      Log("Upgrading all packages from Chocolatey.")
      choco upgrade all -y
    }
    Else {
      Write-Error "Error - Choc ins't in the System... Bye Bye :)"
      Exit
    }
  }

  help {
		Log("usage: all|aws|choco|register|unregister|runnnes|list|update|help")
	}

  refresh {
    Log("Refresh Environment Variables in PowerShell...")
    Get-Refresh-Path
  }

	default {
		Log("usage: all|aws|choco|register|unregister|runnnes|list|update|help")
  }

}
